### PR TITLE
feat!: getting pods

### PR DESCRIPTION
**Pull request by IgorShadurin**
- started using `bee-factory` for integration tests
- implemented getting pods method
- rewritten encryption and decryption algorithm from `crypto` lib to `crypto-js` lib for support not only node, but browsers
- implemented encryption unit tests
- implemented browser tests
- significantly reduced the amount of unused code

### CODE REVIEW

File: src/fairdrive-protocol.ts [hunk 1]
```diff
@@ -39,46 +46,78 @@ export class FairdriveProtocol {
       this.users[username] = address
     } else if (mnemonic) {
       try {
-        this.users[username] = Wallet.fromMnemonic(mnemonic).address
+        const wallet = Wallet.fromMnemonic(mnemonic)
+        this.users[username] = wallet.address
```
Comment (nugaon) [inline src/fairdrive-protocol.ts:50]: why not save the `wallet` itself?

File: src/fairdrive-protocol.ts [hunk 2]
```diff
@@ -39,46 +46,78 @@ export class FairdriveProtocol {
...
     return userInfo
   }
+
+  async podLs(): Promise<Pod[]> {
+    validateActiveAccount(this.accountData)
+
+    const result = await getFeedData(
+      this.accountData.bee,
+      FairdriveProtocol.POD_TOPIC,
+      this.accountData.wallet!.address,
```
Comment (nugaon) [inline src/fairdrive-protocol.ts:106]: that can cause error, can't it? since the wallet is optional maybe it is undefined.

File: src/feed/api.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import { bmtHashString, extractChunkData } from '../account/utils'
+import { getId } from './handler'
+import { bytesToHex } from '../utils/hex'
+import { keccak256Hash } from '../account/encryption'
+import { lookup } from './lookup/linear'
+import { Epoch } from './lookup/epoch'
+import Long from 'long'
+import { NoClue } from './lookup/lookup'
+
+const downloadTimeout = 1000
```
Comment (nugaon) [inline src/feed/api.ts:12]: it has to be an optional parameter (with this or a higher default) in the `getFeedData` function.

File: src/feed/lookup/linear.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import Long from 'long'
+import { Epoch } from './epoch'
+import { Data } from '@ethersphere/bee-js/dist/src/types'
+
+export async function lookup(
+  time: Long,
+  hint: Epoch,
```
Comment (nugaon) [inline src/feed/lookup/linear.ts:7]: not used parameter.

File: src/feed/lookup/linear.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  read: (epoch: Epoch, time: Long) => Promise<Data>,
+): Promise<Data> {
+  if (time.eqz()) {
+    time = Long.fromNumber(Math.round(Date.now() / 1000))
+  }
+
+  let previousChunk
```
Comment (nugaon) [inline src/feed/lookup/linear.ts:14]: missing type

File: src/feed/lookup/linear.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  let level = 31
+  for (;;) {
+    if (level <= 0) {
+      break
+    }
```
Comment (nugaon) [inline src/feed/lookup/linear.ts:19]: why are you not using
```suggestion
  while (level > 0) {
```

File: src/feed/lookup/epoch.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+// import { maxuint64 } from './lookup'
+import { Bytes } from '@ethersphere/bee-js/dist/src/utils/bytes'
+
+const EpochLength = 8
+const MaxTime = Long.fromNumber(1).shiftLeft(56).sub(1)
+
+export declare type EpochID = Bytes<typeof EpochLength>
+
+export function getBaseTime(time: number, level: number): Long {
+  return Long.fromNumber(time).and(Long.MAX_UNSIGNED_VALUE.shiftLeft(level))
```
Comment (nugaon) [inline src/feed/lookup/epoch.ts:11]:
```suggestion
export function getBaseTime(time: Long, level: number): Long {
  return time.and(Long.MAX_UNSIGNED_VALUE.shiftLeft(level))
```

File: src/feed/lookup/epoch.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+export class Epoch {
+  level
+  time
+
+  constructor(level: number, time: Long) {
+    this.level = level
+    this.time = time
+  }
```
Comment (nugaon) [inline src/feed/lookup/epoch.ts:21]:
```suggestion
  constructor(public level: number, public time: Long) {}
```

File: src/feed/lookup/lookup.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import Long from 'long'
+import { Epoch } from './epoch'
+
+export const LowestLevel = 0
+export const HighestLevel = 31
+export const DefaultLevel = HighestLevel
+export const NoClue = new Epoch(0, 0)
+export const worstHint = new Epoch(63, 0)
```
Comment (nugaon) [inline src/feed/lookup/lookup.ts:8]: second parameters are not in types

File: src/feed/lookup/lookup.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+export class Lookup {
+  getNextLevel(last: Epoch, now: Long): number {
+    let mix = last.base().xor(now)
+    mix = mix.or(Long.fromNumber(1).shiftLeft((last.level as number) - 1))
+
+    if (mix.greaterThan(Long.MAX_UNSIGNED_VALUE.shiftRight(Long.fromNumber(64 - HighestLevel - 1)))) {
```
Comment (nugaon) [inline src/feed/lookup/lookup.ts:15]: can you elaborate this line please?

File: src/feed/handler.ts [hunk 11]
```diff
@@ -1,17 +1,18 @@
...
 
 const TopicLength = 32
 
-export function epocId(time: any, level: any): Uint8Array {
-  const result = new Uint8Array(8)
-  result[7] = 31
+export function epocId(time: number, level: number): number[] {
+  const base = Long.fromNumber(time).and(Long.MAX_UNSIGNED_VALUE.shiftLeft(level))
+  const result = base.toBytes(true)
+  result[7] = level
```
Comment (nugaon) [inline src/feed/handler.ts:10]: So does it use only a byte length? then why did you add `long` as a dependency that use 8 bytes? you can use the default bitwise operators on `number` type instead.

File: src/feed/api.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+export async function getFeedData(bee: Bee, topic: string, address: string): Promise<Data> {
+  const addressBytes = Utils.makeEthAddress(address)
+  const topicHash = bmtHashString(topic)
+  const chunk = await lookup(Long.fromNumber(0), NoClue, async (epoch: Epoch, time: Long): Promise<Data> => {
+    const tempId = getId(topicHash, time.toNumber(), epoch.level)
```
Comment (nugaon) [inline src/feed/api.ts:18]: I just don't get you transfer Long into `number` it loses its precision then why are you using `long`?

File: src/feed/lookup/epoch.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+  base(): Long {
+    return getBaseTime(this.time, this.level)
+  }
+
+  id(): EpochID {
+    const base = this.base()
+    const id = Uint8Array.from(base.toBytes()) as EpochID
+    id[7] = this.level
```
Comment (nugaon) [inline src/feed/lookup/epoch.ts:30]: So I guess from this it would use only a bytelength for the epoch serialisation, the other bytes could be only padding for the correct hashing, am I wrong?

**Review by nugaon — CHANGES_REQUESTED** [review]
- use `bee-factory` for integration tests.
- write jsdoc for all exported functions and classes.
- documentation in the README.
- not specific types (e.g. addresses, accounts are both strings, no distinction).
- there are three main changes in this PR that should have been separated into three PRs instead:
	- feat: epoch-based feeds
	- refactor: account and wallet
	- feat: pod listing
- code duplications since the copied codes are also available in the dependency package.
- naming in the feed implementation; that you implemented regards to only the epoch-based feeds and not all feeds as one could assume from the naming.
- only add dependency if it inevitable. We are working with libraries that other projects will use mostly on frontend, so the filesize matters. Moreover, every added package is a potential security issue as well. As I noticed, `long` is not a necessary package as the time operations that you are using it for only use 1 byte instead of 8 bytes that this package provides.

File: src/fairdrive-protocol.ts [hunk 14]
```diff
@@ -22,10 +31,8 @@ export class FairdriveProtocol {
    * @param mnemonic 12 space separated words to initialize wallet
    */
   async userImport(username: string, address = '', mnemonic = ''): Promise<void> {
-    // todo validate address and username
-    if (!username) {
-      throw new Error('Username is required')
-    }
+    validateUsername(username)
+    validateAddress(username)
```
[the remark this answers is not in the document]
Comment (IgorShadurin) [reply src/fairdrive-protocol.ts:35]: everything is right

File: src/feed/handler.ts [hunk 15]
```diff
@@ -1,17 +1,18 @@
 import { Bytes } from '../utils/bytes'
 import { makeContentAddressedChunk } from '../chunk/cac'
+import Long from 'long'
 
 const TopicLength = 32
 
-export function epocId(time: any, level: any): Uint8Array {
-  const result = new Uint8Array(8)
-  result[7] = 31
+export function epocId(time: number, level: number): number[] {
```
[the remark this answers is not in the document]
Comment (IgorShadurin) [reply src/feed/handler.ts:7]: Yes, it's a typo from the original code. Need to fix

File: src/account/utils.ts [hunk 16]
```diff
@@ -17,3 +16,33 @@ export function bmtHashString(stringData: string): Bytes<32> {
...
+  if (!data) {
+    throw new Error('Incorrect username')
+  }
+}
+
+export function validatePassword(data: string): void {
+  if (!data) {
+    throw new Error('Incorrect password')
+  }
+}
```
[the remark this answers is not in the document]
Comment (IgorShadurin) [reply src/account/utils.ts:30]: yes

File: src/feed/lookup/lookup.ts (hunk 10 above)
Comment (IgorShadurin) [reply src/feed/lookup/lookup.ts:15]: I can't explain it because it's translated from the original

File: src/fairdrive-protocol.ts (hunk 1 above)
Comment (IgorShadurin) [reply src/fairdrive-protocol.ts:50]: The principle is taken from the original code. As far as I understand, this is necessary to manage multiple accounts. Not sure if we need support for multiple accounts. On the client side, it is most likely not needed. On the server side, it is needed to fully mimic the behaviour of the current version of FairOS. 
But is it our way? 
This needs to be discussed.

File: src/fairdrive-protocol.ts (hunk 2 above)
Comment (IgorShadurin) [reply src/fairdrive-protocol.ts:106]: The previous line checks that there is an active account and a wallet is present.

File: src/feed/lookup/epoch.ts (hunk 13 above)
Comment (IgorShadurin) [reply src/feed/lookup/epoch.ts:22]: Most likely this is the correct explanation, but the meaning is left to the author of the original golang code.

File: src/feed/api.ts (hunk 12 above)
Comment (IgorShadurin) [reply src/feed/api.ts:17]: When comparing golang long for the first time, it seemed to me that the accuracy might be lost. It might be worth exploring a little more, but taking into account all the other methods (some of which are not yet in our codebase) and their interrelationships.

File: src/account/account-data.ts [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import { Bee, BeeDebug } from '@ethersphere/bee-js'
+import { Wallet } from 'ethers'
+
+export default class AccountData {
+  public readonly bee: Bee
+  public readonly beeDebug: BeeDebug
+  public wallet?: Wallet
```
[the remark this answers is not in the document]
Comment (IgorShadurin) [reply src/account/account-data.ts:7]: This variable is responsible for whether the account has been logged in or not. Some methods (for example, getting a list of pods) pay attention to it.

File: src/feed/api.ts (hunk 12 above)
Comment (nugaon) [reply src/feed/api.ts:17]: I think if it is clear what is exactly stored in the bytearray then clears out the correct handling of the data as well. Now, it seems it is just a timestamp + level

File: src/feed/lookup/linear.ts [hunk 18]
```diff
@@ -0,0 +1,34 @@
+import Long from 'long'
+import { Epoch } from './epoch'
+import { Data } from '@ethersphere/bee-js/dist/src/types'
```
Comment (nugaon) [inline src/feed/lookup/linear.ts:3]: Not found declaration

File: test/integration/fdp-class.browser.spec.ts [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+  beforeAll(async () => {
+    await jestPuppeteer.resetPage()
+    const testPage = join(__dirname, '..', 'testpage', 'testpage.html')
+    await page.goto(`file://${testPage}`)
+  })
+
+  it('should create a new FDP instance in browser', async () => {
+    const urls = await page.evaluate(
+      (BEE_URL, BEE_DEBUG_URL) => {
+        const accountData = new window.FairDataProtocol.FairDataProtocol(BEE_URL, BEE_DEBUG_URL).accountData
```
Comment (nugaon) [inline test/integration/fdp-class.browser.spec.ts:18]: create an import from the FDP's index file in order to get the type, because the IDE does not know about it like this.

File: test/unit/encryption.spec.ts [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+    ]
+
+    it('decrypt', () => {
+      for (const item of examples) {
+        const decrypted = decrypt(item.password, item.data)
+        expect(decrypted).toEqual(item.expected)
+      }
+    })
+
+    it('encrypt', () => {
```
Comment (nugaon) [inline test/unit/encryption.spec.ts:34]: add one more test please, where you encrypt/decrypt the data with the same password and check whether you got the original data.

**Review by nugaon — CHANGES_REQUESTED** [review]
We are slowly getting there, I left some comments below and my additional experiences are:
- I get warning message at importing the library with node v16: `WARNING: Missing strong random number source`. It is a huge red flag as the library generate keys.
- There are some missing type definitions e.g. at `TextEncoder`
- `long` library is still a dependency after we discussed it may be not necessary.
- update the repository's README with introducing new features and giving code snippets for that.
- remove `debug-server` as it is not used.
- some functions/ts files are not in the best fitting library such as `chunk/seralise.ts` could be moved to `utils/bytes.ts` as it is a bytes related function.
- you have started to use classes but the data/code is not in one place and also the methods/fields are not quite intuitive. For that, I suggest the following structure (references to the corresponding namings are after the arrows):
```
FairDataProtocol {
  Account (AccountData) {
    wallet [set; get] <- activeAccount
    usernameToAddreess <- users
    import(userName, mnemonic) <- userImport
    login() <- userLogin
    register <- userSignUp
  }
  PersonalStorage <- new class for POD related functionalities {
    list <- PodLs
    const POD_TOPIC
  }
}
```
In regard to the new `PersonalStorage` class please create a separate folder for it with name `pod` and place the interface `Pod` there as well.

File: src/index.ts [hunk 21]
```diff
@@ -1,11 +1,11 @@
...
 
 // for require-like imports
 declare global {
   interface Window {
-    // bound as 'FairdriveProtocol' via Webpack
-    FairdriveProtocol: {
-      FairdriveProtocol: typeof import('./fairdrive-protocol').FairdriveProtocol
+    // bound as 'FairDataProtocol' via Webpack
+    FairDataProtocol: {
+      FairDataProtocol: typeof import('./fair-data-protocol').FairDataProtocol
```
[the remark this answers is not in the document]
Comment (IgorShadurin) [reply src/index.ts:8]: I agree, it's a bit weird.

File: README.md [hunk 22]
```diff
@@ -1,3 +1,158 @@
...
+
+Install project dependencies with
+
+```sh
+npm i
+```
+
+### Test
+
+The tests run in both context: node and dom with Jest.
```
Comment (nugaon) [inline README.md:113]: Jest and Puppeteer

File: README.md [hunk 23]
```diff
@@ -1,3 +1,158 @@
...
+
+To run the integration tests, you need to use our [`bee-factory`](https://github.com/fairDataSociety/bee-factory/) project. Clone the repo, you can use our prebuilt Docker images with setting env. variables like:
+
+```bash
+export FAIROS_VERSION="0.7.3"
+export BLOCKCHAIN_VERSION="1.2.0"
+export BEE_ENV_PREFIX="swarm-test"
+export BEE_IMAGE_PREFIX="docker.pkg.github.com/ethersphere/bee-factory"
+```
```
Comment (nugaon) [inline README.md:122]: These are default env variables that the project contains. I wouldn't be so verbose about it, for more detail they can check out FDP's bee-factory repo

File: README.md [hunk 24]
```diff
@@ -1,3 +1,158 @@
...
+
+Customize these values based on which FairOS version you want to run. After the env. variables are set use the `./scripts/environment.sh` script with `start --fairos` parameter.
+
+If you want to skip creation of postage stamps every run of integration tests you can create stamps for both nodes and set them under env. variables `BEE_POSTAGE` and `BEE_PEER_POSTAGE`.
+
+By default, for integration tests two bee nodes are expected to run on localhost on addresses `http://localhost:1633` and `http://localhost:1635`. These are the default values for the `bee-factory` script.
+If you want to use custom setup, you can change the behavior of tests to different addresses using environment variables `BEE_API_URL`, `BEE_DEBUG_API_URL`, `BEE_PEER_DEBUG_API_URL` and `BEE_PEER_API_URL`.
+
+In Visual Studio environment, the tests have been set up to run against your local bee node on `http://localhost:1633`
+To run Jest tests, choose the `vscode-jest-tests` CI job under the Run tab.
```
Comment (nugaon) [inline README.md:132]: The configuration files are not defined for this.
Anyway, I would remove this section, we don't need it

File: src/account/account-data.ts [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  /**
+   * Logs in with the FDP credentials and gives back ethers wallet
+   *
+   * @param username FDP username
+   * @param password password of the wallet
+   * @returns BIP-039 + BIP-044 Wallet
+   */
+  async login(username: string, password: string): Promise<Wallet> {
```
Comment (nugaon) [inline src/account/account-data.ts:72]: I would add an optional parameter for `address` that falls back to `this.usernameToAddress` in order to login in one line

**Review by nugaon — CHANGES_REQUESTED** [review]
This batch of work is the best so far! Congrats.

The encryption does not work unfortunately with node v17, for that please create an issue and fix it in an other PR.

I had some change requests below, but after then I think we can merge : )

**Review by nugaon — APPROVED** [review]
LGTM! : )
